### PR TITLE
feat(stream-json): Add JSON stream adapters

### PR DESCRIPTION
**Pull request by kriskowal**
This builds on #964, providing adapters for Uint8Array message streams to JSON streams, capturing the concerns of transcoding and marshaling. These are also hardened.

### CODE REVIEW

File: packages/stream-json/reader.js
Comment (mhofman) [inline packages/stream-json/reader.js:21]: Afaik `JSON.parse` is types to return `any`, right?

It might be more useful(but possibly unsafe) to provide a generic version here.

File: packages/stream-json/writer.js
Comment (mhofman) [inline packages/stream-json/writer.js:12]: Any reason not to write this as generator like reader?

Comment (kriskowal) [reply packages/stream-json/writer.js:12]: Yeah, theoretically there are two kinds of generators: even and odd. An odd generator starts at the top of the function and even generator would advance immediately to the first yield. We only have odd generators, which means that the first value sent to a generator gets discarded. It’s possible to compensate by using a wrapped generator and inducing `next()` immediately, but the type is hard to pin down.

File: packages/stream-json/reader.js
Comment (kriskowal) [reply packages/stream-json/reader.js:21]: I was not so sure because the signature is theoretically narrower than that. I figured stating this would give the types room to evolve.

File: packages/stream-json/reader.js
Comment (erights) [inline packages/stream-json/reader.js:5]:
```suggestion
 * writer streams are modeled as hybrid async iterators + generators.
```

File: packages/stream-json/reader.js
Comment (erights) [inline packages/stream-json/reader.js:17]: It is a shame we need to leave Jessie. Perhaps we can come up with a design rule for restricted async generators like we have for async functions. For example, if the `yield` is always at the top level of the body of an top level for-await-of, that would cover this case, probably many other related cases, and would still admit a simple implementation strategy short of CPS transforming.

File: packages/stream-json/reader.js
Comment (erights) [inline packages/stream-json/reader.js:1]: Huh. I just noticed. Are we not following the `src` vs `test` directory convention? I think I like that better than having a bunch of source files at the package level, but I have not thought about it.

File: packages/stream-json/writer.js
Comment (erights) [reply packages/stream-json/writer.js:12]: The old `function.next` tc39 proposal died because no one was willing to invest to move it forward. That would have allowed generators (and async generators) to read the argument of the first `next` call. Oh well.

I'll use this as an opportunity to lobby, separately, for an updater adaptor which would also rename the API surface. But these PRs should not block on that debate.

**Review by erights — APPROVED** [review]
LGTM

File: packages/stream-json/reader.js
Comment (kriskowal) [reply packages/stream-json/reader.js:1]: The rule is generally the same, but I am proposing an amendment and I think I have @michaelfig’s blessing:

With Node.js implementing ESM they added `imports` and `exports` features to `package.json` that make it possible to expose more than `main` as the package’s exported API surface, and this is now *enforced*. But, to make a package that straddles the old world and the new, you have to keep the `.js` extension.

It has always been a faux pas to reach like `import "@org/package/src/internal.js`. The `src` directory should indicate that the modules within are not participating in the public API. It’s more proper to `import '@netstring/reader.js'`. It is possible with an `exports` directive to even make this `import '@netstring/reader'`, just not portably.

So I’m proposing that we surface all public API modules at the top-level. It would certainly be consistent with our existing rule to require each of these a thunk for an internal module, e.g., `export * from './src/internal.js'`, but I feel that would be wasteful and fragile.

**Comment by kriskowal** [comment]
@erights What API and behavior do you propose for an updater?

File: packages/stream-json/reader.js
Comment (kriskowal) [reply packages/stream-json/reader.js:17]: We certainly could implement this particular case with the iterator interface directly and avoid the need, but there are certainly other kinds of adapters that benefit from async generators.

**Pull request merged**
